clients/geth/specular/rollup/services/sequencer/sequencer.go

```diff
@@ -62,37 +62,138 @@ func New(eth services.Backend, proofBackend proof.Backend, cfg *services.Config,
+	// Check if tx in batch
+	txIndex := getTxIndexInBatch(batchTxs, tx)
+	if txIndex >= 0 {
+		// Remove if exists
+		batchTxs = append(batchTxs[:txIndex], batchTxs[txIndex+1:]...)
+	} else {
+		// Check if tx exists on chain
+		prevTx, _, _, _, err := s.ProofBackend.GetTransaction(s.Ctx, tx.Hash())
+		if err != nil {
+			log.Error("Checking GetTransaction, this is err", "error", err)
+			return batchTxs
+		}
+		if prevTx != nil {
+			// Remove prexTx if exists
+			txIndex = getTxIndexInBatch(batchTxs, tx)
+			if txIndex >= 0 {
+				batchTxs = append(batchTxs[:txIndex], batchTxs[txIndex+1:]...)
+			}
+		} else {
+			batchTxs = append(batchTxs, tx)
+		}
+	}
+	return batchTxs
+}
+
+// Process sorted txs
+func (s *Sequencer) processSortedTxs(sortedTxs *types.TransactionsByPriceAndNonce, batcher *Batcher, batchTxs []*types.Transaction) {
+	if sortedTxs == nil {
+		log.Info("processSortedTxs -> sortedTxs is nil")
+	}
+	for {
+		tx := sortedTxs.Peek()
+		if tx == nil {
+			break
+		}
+		// Check if tx in batchTxs
+		batchTxs = s.modifyTxnsInBatch(batchTxs, tx)
+		sortedTxs.Pop()
+	}
+	if len(batchTxs) == 0 {
+		return
+	}
+	err := batcher.CommitTransactions(batchTxs)
+	if err != nil {
+		log.Crit("Failed to commit transactions", "err", err)
+	}
+	blocks, err := batcher.Batch()
+	if err != nil {
+		log.Crit("Failed to batch blocks", "err", err)
+	}
+	batch := rollupTypes.NewTxBatch(blocks, 0) // TODO: add max batch size
+	s.batchCh <- batch
+}
+
 // This goroutine fetches txs from txpool and batches them
-// TODO: fetching txs through sub is unstable
-// TODO: batch txs by timer and directly querying txpool as worker does
 func (s *Sequencer) batchingLoop() {
 	defer s.Wg.Done()
 	defer close(s.batchCh)
 
+	// Ticker
+	var ticker = time.NewTicker(timeInterval)
+	defer ticker.Stop()
+
 	// Watch transactions in TxPool
 	txsCh := make(chan core.NewTxsEvent, 4096)
 	txsSub := s.Eth.TxPool().SubscribeNewTxsEvent(txsCh)
 	defer txsSub.Unsubscribe()
 
+	// Process txns via batcher
 	batcher, err := NewBatcher(s.Config.Coinbase, s.Eth)
 	if err != nil {
 		log.Crit("Failed to start batcher", "err", err)
 	}
 
+	var batchTxs []*types.Transaction
+
+	// Loop over txns
 	for {
 		select {
+		case <-ticker.C:
+			// Get pending txs - locals and remotes, sorted by price
+			var txs []*types.Transaction
+			var sortedTxs *types.TransactionsByPriceAndNonce
+			signer := types.MakeSigner(batcher.chainConfig, batcher.header.Number)
+
+			pending := s.Eth.TxPool().Pending(true)
+			localTxs, remoteTxs := make(map[common.Address]types.Transactions), pending
+			for _, account := range s.Eth.TxPool().Locals() {
+				if txs = remoteTxs[account]; len(txs) > 0 {
+					delete(remoteTxs, account)
+					localTxs[account] = txs
+				}
+			}
+			if len(localTxs) > 0 {
+				sortedTxs = types.NewTransactionsByPriceAndNonce(signer, localTxs, batcher.header.BaseFee)
+				s.processSortedTxs(sortedTxs, batcher, batchTxs)
```

Comment:
you send two batches in one tick, and also in this way you have to filter some redundant txs between local/remote by calling the chain API instead of directly comparing them to the local array.